### PR TITLE
LPS-195732 Merge BrowserSniffer interface and impl logic into Util in kernel

### PR DESCRIPTION
**Pull request by julschong**
resubmitted from [LPS-195732 Merge BrowserSniffer interface and impl logic into Util in kernel](https://github.com/liferay-core-infra/liferay-portal/pull/5519#top)

### CODE REVIEW

File: portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java [hunk 1]
```diff
@@ -28,12 +26,7 @@
  * @author Shuyang Zhou
  * @author Iliyan Peych
  */
-public class BrowserSnifferImplTest {
-
-	@ClassRule
-	@Rule
-	public static final LiferayUnitTestRule liferayUnitTestRule =
-		LiferayUnitTestRule.INSTANCE;
```
Comment (julschong) [inline portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java:35]: changed file name and move to kernel

File: portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java [hunk 2]
```diff
@@ -33,22 +33,6 @@ public class BrowserSnifferUtil {
...
-
-	public static final String[] VERSION_LEADINGS = {
-		"edge", "chrome", "firefox", "version", "minefield", "trident"
-	};
-
-	public static final char[] VERSION_SEPARATORS = {
-		CharPool.BACK_SLASH, CharPool.SLASH
-	};
-
 	public static boolean acceptsGzip(HttpServletRequest httpServletRequest) {
```
Comment (julschong) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java:36]: change to private

File: portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java [hunk 3]
```diff
@@ -157,8 +141,8 @@ public static String getRevision(HttpServletRequest httpServletRequest) {
 		}
 
 		revision = parseVersion(
-			_getUserAgent(httpServletRequest), REVISION_LEADINGS,
-			REVISION_SEPARATORS);
+			_getUserAgent(httpServletRequest), _REVISION_LEADINGS,
+			_REVISION_SEPARATORS);
```
Comment (julschong) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java:146]: use reflection

File: portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java [hunk 4]
```diff
@@ -190,11 +174,12 @@ public static String getVersion(HttpServletRequest httpServletRequest) {
 
 		String userAgent = _getUserAgent(httpServletRequest);
 
-		version = parseVersion(userAgent, VERSION_LEADINGS, VERSION_SEPARATORS);
+		version = parseVersion(
+			userAgent, _VERSION_LEADINGS, _VERSION_SEPARATORS);
 
 		if (version.isEmpty()) {
 			version = parseVersion(
-				userAgent, REVISION_LEADINGS, REVISION_SEPARATORS);
```
Comment (julschong) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java]: use reflection

File: portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java [hunk 5]
```diff
@@ -462,4 +447,20 @@ private static String _getUserAgent(HttpServletRequest httpServletRequest) {
...
+	};
+
+	private static final String[] _VERSION_LEADINGS = {
+		"edge", "chrome", "firefox", "version", "minefield", "trident"
+	};
+
+	private static final char[] _VERSION_SEPARATORS = {
+		CharPool.BACK_SLASH, CharPool.SLASH
+	};
+
```
Comment (julschong) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/BrowserSnifferUtil.java:465]: changed to private

File: portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java [hunk 6]
```diff
@@ -140,14 +138,20 @@ public void testParseVersion() throws IOException {
 				Assert.assertEquals(
 					parts[0].trim() + " version", parts[1].trim(),
 					BrowserSnifferUtil.parseVersion(
-						userAgent, BrowserSnifferUtil.VERSION_LEADINGS,
-						BrowserSnifferUtil.VERSION_SEPARATORS));
+						userAgent,
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_VERSION_LEADINGS"),
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_VERSION_SEPARATORS")));
```
Comment (julschong) [inline portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java:146]: use reflection

File: portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java [hunk 7]
```diff
@@ -140,14 +138,20 @@ public void testParseVersion() throws IOException {
...
 				Assert.assertEquals(
 					parts[0].trim() + " revision", parts[2].trim(),
 					BrowserSnifferUtil.parseVersion(
-						userAgent, BrowserSnifferUtil.REVISION_LEADINGS,
-						BrowserSnifferUtil.REVISION_SEPARATORS));
+						userAgent,
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_REVISION_LEADINGS"),
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_REVISION_SEPARATORS")));
```
Comment (julschong) [inline portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java:154]: use reflection

File: portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java [hunk 8]
```diff
@@ -140,14 +138,20 @@ public void testParseVersion() throws IOException {
 				Assert.assertEquals(
 					parts[0].trim() + " version", parts[1].trim(),
 					BrowserSnifferUtil.parseVersion(
-						userAgent, BrowserSnifferUtil.VERSION_LEADINGS,
-						BrowserSnifferUtil.VERSION_SEPARATORS));
+						userAgent,
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_VERSION_LEADINGS"),
+						ReflectionTestUtil.getFieldValue(
+							BrowserSnifferUtil.class, "_VERSION_SEPARATORS")));
```
Comment (julschong) [inline portal-kernel/test/unit/com/liferay/portal/kernel/servlet/BrowserSnifferUtilTest.java:145]: use reflection

**Comment by Kvale1733** [comment]
@julschong No regressions found. Need to rebase.

**Comment by julschong** [comment]
rebased

**Comment by julschong** [comment]
ci:test:sf

**Comment by julschong** [comment]
@shuyangzhou, not moving this Util into portal-impl

```
Following build.gradles do not contain portal-impl dependencies: 
[ "./modules/apps/social/social-bookmarks-taglib/build.gradle", "./modules/apps/portal-url-builder/portal-url-builder-impl/build.gradle"
]
```

**Comment by julschong** [comment]
closed per our discussion

**Pull request closed without merging**